src/engine/tools.cpp

```diff
@@ -395,17 +400,12 @@ namespace fheroes2
 
     uint32_t calculateCRC32( const uint8_t * data, const size_t length )
     {
-        uint32_t crc = 0xFFFFFFFF;
-        for ( size_t i = 0; i < length; ++i ) {
-            crc ^= static_cast<uint32_t>( data[i] );
-
-            for ( int bit = 0; bit < 8; ++bit ) {
-                const uint32_t poly = ( crc & 1 ) ? 0xEDB88320 : 0x0;
-                crc = ( crc >> 1 ) ^ poly;
-            }
+        if ( length > std::numeric_limits<uInt>::max() ) {
+            throw std::
+                system_error( std::make_error_code( std::errc::value_too_large ),
+                              "Too large `length` provided to `calculateCRC32`. Must be no larger than `std::numeric_limits<uInt>::max()` (usually `(1 << 32) - 1`)." );
         }
-
-        return ~crc;
+        return ~crc32( 0, data, static_cast<uInt>( length ) );
```

Comment:
I tested this PR with Russian (Buka) assets.
The result of CRC32 calculation now is 0x2A30750C (707818764), in binary:
00101010001100000111010100001100
and should be: 0xD5CF8AF3 (3587148531), in binary:
11010101110011111000101011110011

As I can conclude now we don't need to invert the bits of returned value:
```suggestion
        return crc32( 0, data, static_cast<uInt>( length ) );
```